Plane to use AHRS_MAX_WIND as documented

ChrisBird opened the pull request:
This is to provide an initial response to the issue raised in #9372 that we currently dont use MAX_WIND as documented.

This is not the end solution for this problem, I'll be working on a longer term solution in the coming weeks but this is the first item to at least use something as documented.

This has been tested by modifying SIM_Aircraft to use:
+    //airspeed_pitot = constrain_float(velocity_air_bf * Vector3f(1.0f, 0.0f, 0.0f), 0.0f, 120.0f);
+    airspeed_pitot = constrain_float(velocity_air_bf * Vector3f(1.0f, 0.0f, 0.0f), 0.0f, 12.0f + rand_float());
That mod should lock the airspeed to a value of 11 to 13m/s.

Can someone please check my logic that I haven't messed something up somewhere.

This PR should remain open for a week or so as I'll be adding further code to reset the counter to 0 after a period of no bad readings (failing the wind_max threshold test from gnd speed).

Currently I have it set to 30 which should be 3 seconds, this can be shortened or lengthened.  Those that have experienced this issue can provide guidance on how long a problem it can be before we need to disable it.  Some have suggested 4 seconds.

ChrisBird commented:
I'll need to think about how I handle multiple airspeed sensors, currently this only addresses the primary airspeed sensor.

IamPete1 commented:
I not sure what people would set AHRS_WIND_MAX as, but I guess to be most effective it should be fairly close to maximum expected wind speeds. Maybe this should to re-enable the airspeed sensor if the wind-speed reading drops bellow AHRS_WIND_MAX again? So as not to be permanently disabled in a gust, but then again that could cause some funny behaviour if the sensor has genuinely failed and it just outputting nonsense.

ChrisBird commented:
It puts out a gcs warning message: "Airspeed sensor - possibly faulty - ARSPED_USE: 0"

ChrisBird commented:
@IamPete1 From memory its positive and the ground course would be reversed.  I'll need to check that though. Good pickup.

lowlyelevated commented:
How does this interact with airspeed blending and dual airspeed sensors? If possible, it'd be useful to have it monitor each sensor independently, so if one fails and outputs an unrealistic value, it can default to the other sensor instead of disabling both.

Evan1010 commented:
Specifically, @lowlyelevated is referring to [this ](https://github.com/ArduPilot/ardupilot/pull/7738)merged pull request, which uses healthy/ not healthy flags for multiple airspeed sensors to switch between them. What's unclear to me is whether it will switch to synthetic airspeed if both are marked unhealthy . . .

EShamaev commented:
Ground speed and wind speed are vectors. I doubt that simple addition can be used with respect to airspeed comparison.

WickedShell commented:
IMO the entire idea of `AHRS_WIND_MAX` only works if you set a very large floor value. I've seen people be wrong at what the winds are aloft far more often, in which case you would want to set a very conservative estimate (IE something that is very far outside of plausible, so if you would be happy in flying upto 13 m/s wind, then the limit should be set up at 20 m/s or more). This  only becomes useful once it's extremely far outside the tuned envelope. In my opinion we should instead be assessing if the wind estimation we have is physically consistent with what the aircraft is observing from the sensors and looking for a divergence.

@ChrisBird as a general note using `rand()` for wind simulation via the airspeed reading is only a test of our ability to reject pseduorandom noise through the airspeed sensor. It doesn't allow the vehicle to get a decent wind simulation or estimation as it is not physically accurate. (That said a small low level of noise is realistic). As far as testing in SITL you should have been able to test this by simply increasing or decreasing the wind in the simulation. If you want to introduce error into the sensor in flight the best way is to simply modulate the ARSPD_OFFSET parameter as we simulate an analog sensor in SITL.

WickedShell commented:
> Specifically, @lowlyelevated is referring to this merged pull request, which uses healthy/ not healthy flags for multiple airspeed sensors to switch between them. What's unclear to me is whether it will switch to synthetic airspeed if both are marked unhealthy . . .

To the best of my knowledge that was reverted. Or if it wasn't it has under seen really heavy refactoring since then and I can no longer find it in the code.

ChrisBird commented:
@EShamaev gps ground speed is a float, there is also ground course which gives it a direction.

ChrisBird commented:
@WickedShell i agree on the setting of wind_max.  In the case of one log i reviewed it reported values between 16 and 19m/s while having a ground speed of over 40m/s, throttle maxed and it was evident from the pattern it was flying that the wind was at most a few m/s.  I'd love to review more logs with this issue.  Agreed on the better solution, thats why ive said this is not the end solution.  Will have a play with the sitl suggestion, can be auto tested then.

ChrisBird commented:
I think we need a simple but rough solution and then a more precise method later on.  People have crashed planes when it is avoidable.  The idea is to get the parameter to behave as the documentation says.  Then the better more precise solution can be implemented.  Key is to ensure this change isnt detrimental to current airframes.

Evan1010 commented:
I didn't even notice that it was reverted and in rework. Thank you for pointing that out.

I guess my opinion on this depends on timeline and functionality:
1. If #7891 is going to be an ongoing development for quite some time, I do think a quick fix is in order.
2. If #7891 addresses the same failure mode(s) as this pull request, then perhaps it can replace it when ready?
3. If this pull request addresses a failure mode which #7891 does not (I am making no assumptions), then it can either stand on its own or be an additional case addressed  by #7891.

I think point number 2 is the most likely case from what I've read. Completely independent of these points however, is that [AHRS_WIND_MAX](http://ardupilot.org/plane/docs/parameters.html#ahrs-wind-max-maximum-wind) does not currently function as intended. @EShamaev Do you think that the functionality as ~described~ implied in AHRS_WIND_MAX is undesirable in the long term? It doesn't prescribe a method for coping with the failing airspeed sensor, but just says that it does.

ChrisBird commented:
@Evan1010 thank you for that, its basically how i view it too.  Do you have some logs that my help, ive only got a handful and they seem to fail in the same way.  Locked around a value with some minor noise.  Ive had the issue but luckily i had arspd_use at 0 as i was testing.

Evan1010 commented:
I'll try to dig up some logs - they're from a while ago, after all.

@EShamaev I don't want to keep bothering you, but do you think the uncertainty in airspeed direction given by a failed sensor can be overcome for a more proper comparison between ground speed, air speed, and wind? As it is, I believe the code effectively assumes the worst case scenarios of a direct tailwind for the first comparison case (aspeed > (gnd_speed + wind_max)), and a direct headwind for the second comparison case (aspeed < (gnd_speed - wind_max)).

Naterater commented on ArduPlane/sensors.cpp:
```diff
@@ -77,6 +79,27 @@ void Plane::read_airspeed(void)
+                airspeed_failed_count++;
+
+                if(airspeed_failed_count > 30) {
+                    gcs().send_text(MAV_SEVERITY_WARNING, "Airspeed sensor - possibly faulty - ARSPED_USE: 0");
```
I think the warning should be a little different. Thoughts on

"Airspeed fault - AHRS_WIND_MAX exceeded - airspeed use disabled"

ChrisBird replied:
Agreed I'll make change.

ChrisBird commented:
Hi @Naterater it is pulling it from the parameter.  Where do you see it pulling it from ekf?

Naterater commented:
Sorry, I'm unfamiliar with helper functions but I'm happy that's what's going on. Will edit my post above again, please excuse the ignorance.

ChrisBird commented:
I have added the auto recovery check after a period of time.  Happy to change the figures over to '#define' instead so its better documentation wise.  Will do that over the weekend.

I've done quite a bit of SITL testing.  I've discovered there is a SIM_ARSPD_FAIL parameter too, as well as the other SIM_WIND_ methods that people have suggested.  If we put a plane up and the operator gets the wind estimate wrong, it should seamlessly swtich between the two.  A good way of knowing if you know your wind speeds I guess.  The question is could this cause a plane to come down?

So what I've done is I've made this use of the airspeed a type 3 use.  I've had to rework a bunch of the existing functions to work but it didnt after doing some digging nothing else is using these functions.... Or they are used via other help functions which now account for the change.

This way if you want to have this mode of checking you can, otherwise it uses the normal throttle up and plough into the ground if your airspeed sensor fails.  Having said that in testing I've had more altitude gains than ploughing into the ground......

One bug I'm tracking down, the recovery period seems to be quicker than the 9 seconds I've got it listed for, although I've not timed it specifically.  Will be testing that some more.

I'll add a autotest case and test multiple airspeed sensors (I have no idea how this will behave) over the weekend.

Again theres some additional logic, please check that I've got this right - it looks like I do with the SITL testing.....

I'll come back to a longer term solution in a month or two unless there is a better one in the pipeline that will get across the line in the next 2 or 3 months?  There a number of PR's that could meet that, so sing out if you intend to finish them off, or if not but you think the existing PR is a good starting point.

Still looking for logs for when this occurs, basically I've got only the failure mode that @Naterater has and myself.  Would really like to see more logs of this happening - I want to make sure we cover off all the failure modes.

tridge commented on ArduPlane/sensors.cpp:
```diff
@@ -65,6 +65,10 @@ void Plane::accel_cal_update() {
  */
 void Plane::read_airspeed(void)
 {
+    static int airspeed_failed_count=0;
```
please put variables in a structure in the Plane class, instead of using statics

ChrisBird replied:
Ended up moving them to the airspeed state, so they are not tied to the vehicle code, it keeps them contained in the airspeed library.

tridge commented on ArduPlane/sensors.cpp:
```diff
@@ -77,6 +81,45 @@ void Plane::read_airspeed(void)
     }
+
+    float wind_max = ahrs.get_max_wind();
+    if(ahrs.airspeed_sensor_enabled() && airspeed.use() == 3) {
```
this logic should be split out into a separate function, maybe check_airspeed_health()

ChrisBird replied:
Done and moved to the airspeed library

tridge commented on libraries/AP_Airspeed/AP_Airspeed.cpp:
```diff
@@ -458,18 +458,18 @@ void AP_Airspeed::setHIL(float airspeed, float diff_pressure, float temperature)
 }
 
-bool AP_Airspeed::use(uint8_t i) const
+int AP_Airspeed::use(uint8_t i) const
```
if this now takes on more than true/false then it should return an enum

ChrisBird replied:
Returned it back to returning a boolean in the rejigged code.

tridge commented on libraries/AP_Logger/LogFile.cpp:
```diff
@@ -1534,7 +1534,7 @@ void AP_Logger::Write_Airspeed(AP_Airspeed &airspeed)
             rawpressure   : airspeed.get_corrected_pressure(i),
             offset        : airspeed.get_offset(i),
-            use           : airspeed.use(i),
+            use           : (bool)airspeed.use(i),
```
why cast to bool here?

ChrisBird replied:
It has been returned back to returning a boolean so this is no longer required.  But I think there was an underlying issue in that I dont understand why I needed to cast it to a bool.  Surely it would have picked up the change in size automatically?  Or did I miss something there?

tridge commented on libraries/AP_Airspeed/AP_Airspeed.h:
```diff
@@ -78,8 +78,16 @@ class AP_Airspeed
-    bool use(uint8_t i) const;
-    bool use(void) const { return use(primary); }
+    int use(uint8_t i) const;
+    void set_use(int newstate) {
```
why doesn't this just do param[primary].use = newstate?
even better to use set_and_notify, so the GCS gets the update

ChrisBird replied:
I will update it to use set_and_notify as I've noticed that the its annoying when the GCS is not updated with this change in state.

ChrisBird replied:
Changed it to use the set_and_notify, thanks for that suggestion.  Far easier to test when the gcs has the right parameter values.

ChrisBird commented on ArduPlane/sensors.cpp:
```diff
@@ -65,6 +65,10 @@ void Plane::accel_cal_update() {
  */
 void Plane::read_airspeed(void)
 {
+    static int airspeed_failed_count=0;
```
@WickedShell its not only about the time but the fact that we've built up a number of failures, it just happens that they are 3 seconds worth of failures.  If there is a way to know the rate that it's getting called?  If so I'd rather keep the count, but have the check of the counter against 3 seconds worth of failures if that makes sense.  Note this will be moving from sensors.cpp as per the dev call.

magicrub replied:
static variables are not allowed. If The Plane class was ever created twice it would called problems. We don't support that now, but it may be an option in the future.. so best to do it right just in case. Also, never use counters. Always use time based events.

ChrisBird replied:
Remove the static's and put the counters in the airspeed state instead.  I absolutely need the counters, I need to know when we have accumulated enough errors.  It needs to build up a number of errors over a short duration.  I'll add a clearing feature so that every minute or two it clears the counters, to avoid a build up over a long period of time.

magicrub replied:
again, if we decide to change this to 10 or 50Hz or 400Hz then the algorithm is hozed and it's difficult to track down. Try a low-pass filter on a probability like the is_flying flag
https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/is_flying.cpp#L130

ChrisBird replied:
I like that method using the low pass filter, so I then just need to set the desired percentage for the failure and recovery.  Thank you @magicrub that is a much better way of doing it.

ChrisBird commented on libraries/AP_Airspeed/AP_Airspeed.h:
```diff
@@ -78,8 +78,16 @@ class AP_Airspeed
-    bool use(void) const { return use(primary); }
+    int use(uint8_t i) const;
+    void set_use(int newstate) {
+        if(newstate) {
```
Returned this back to how it was before, no need for it anymore as we dont use ARSPD_USE with a value of 3 anymore.

ChrisBird commented on ArduPlane/sensors.cpp:
```diff
@@ -77,6 +81,45 @@ void Plane::read_airspeed(void)
+                // Roughly 9 seconds of clear readings need to be observed before it will clear the airspeed use)
+                if(airspeed_success_count > 90) {
+                    gcs().send_text(MAV_SEVERITY_WARNING, "Airspeed fault cleared- airspeed use enabled");
+                    airspeed.set_use(3);
```
I've removed the ARSPD_USE value of 3, so no longer required.  I had documented it but I did it after that push.

ChrisBird commented:
Just need to check that people are happy with the multiple airspeed sensor logic.  If either of them trip a failure then it will switch to sync airspeed (use=0) for that sensor only.  Only the sensor that tripped the issue can resolve it by itself.  So if the first one fails but the second is still getting good readings I think this will use the good sensor still.

Of course that is reliant on having an appropriate AHRS_WIND_MAX setting.  Added a check that it needs to be greater than 0 to use it.  So the option must be on and the AHRS_WIND_MAX must be set to greater than 0.

I've added a counter reset, that is every 60 seconds it resets the counters.  So to trip a failure you need 5 seconds of failures in a 60 period.  To clear you require 30 seconds of good readings to enable its use again.  This avoids it flipping between them, the last thing you want is for it to 'clear' the faulty sensor as you come in to land.

Made all the time values a #define in airspeed.h so that its clearer.

It's now an option that can be turned on or off.  I envisage I'll work on the estimation method discussed on the dev call and that will become another method.  The options bitmask added is called ARSPD_OPTIONS, note there is only one - its not per sensor.  Also with its name it could be used for 15 other things.  If we want it dropped to 8 bits I'm happy to do that too.

Please check what I've done and if there are issues let me know.

ChrisBird commented:
@tridge can you check that what I've implemented is close to what was discussed on the dev call.

ChrisBird commented:
I'm also not sure i like the reenabling of the sensor.  Thinking of the failure scenario that i have logs for it could be enabled as it slowed for landing, resulting in rapid acceleration.

I think I'll add an option to renable it.  That way a user can choose if they want that.

What are peoples thoughts.

Naterater commented:
Do not re-enable it. Too risky. Flying with a crosswind will likely cause re-enabling.

auturgy commented:
I disagree. If the sensor is faulty, or incorrectly installed, the error will persist and it won’t clear.
If it’s a transient, like moisture, it could, and should be permitted to once testing demonstrates the method works.

Regards,

James 

> On 28 Jan 2019, at 1:40 pm, Nathan E <notifications@github.com> wrote:
> 
> Do not re-enable it. Too risky. Flying with a crosswind will likely cause re-enabling.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

ChrisBird commented:
Hi @Naterater and @auturgy

Thinking about it the reenable option has a place, the water droplet in the pitot tube that clears (As indicated on the dev call, I'd like to see some logs of that scenario......)  I'd only turn it on if I was going to be flying in inclement weather (which most civil RPA / UAS dont normally do) and unlikely i'd want it on during landing.  By default they'll all be off even the basic check, so people can choose..

magicrub commented on libraries/AP_Airspeed/AP_Airspeed.h:
```diff
@@ -9,6 +9,15 @@
+#define AP_AIRSPEED_FAILURE_NUM_CLEAR        120     // 120 / 4 = 30 seconds of good readings estimates in 60 secs
+
+// Bitmask for airspeed options
+#define AP_AIRSPEED_FAILURE_MASK_BASIC       (1<<0)  // If set then use basic airspeed failure check
```
These defines should all start with AP_AIRSPEED_OPTIONS_ and AP_AIRSPEED_OPTIONS_DEFAULT needs to be with it

magicrub commented on libraries/AP_Airspeed/AP_Airspeed.h:
```diff
@@ -213,17 +226,26 @@ class AP_Airspeed
+        uint8_t airspeed_failure_basic_success_count=0;
+        bool    airspeed_failure_basic_renable=false;
+        uint32_t airspeed_failure_last_check=0;
+        uint32_t airspeed_failure_last_reset=0;
```
remove the assignments, the default is already 0 but since they are usually in an array you're better off doing this in init() if it's really needed. Also, if they're all airspeed_failure then it might be easier to group them into a failure struct and drop the airspeed_failure... name.

magicrub commented on libraries/AP_Airspeed/AP_Airspeed.h:
```diff
@@ -213,17 +226,26 @@ class AP_Airspeed
     float get_pressure(uint8_t i);
     void update_calibration(uint8_t i, float raw_pressure);
     void update_calibration(uint8_t i, const Vector3f &vground, int16_t max_airspeed_allowed_during_cal);
+    bool check_airspeed_health_basic(uint8_t i);
```
I'm not sure I like the word "basic" throughout all of this. Either way, rename to check_health_basic. You don't need to name things airspeed in the airspeed library

ChrisBird replied:
Sure no probs

magicrub commented on libraries/AP_Airspeed/AP_Airspeed.h:
```diff
@@ -213,17 +226,26 @@ class AP_Airspeed
         Airspeed_Calibration calibration;
         float last_saved_ratio;
         uint8_t counter;
+        uint8_t reenable_use;
```
reenable_use should signed to match AP_Int8 use

ChrisBird replied:
Sure no probs

magicrub commented on libraries/AP_Airspeed/AP_Airspeed_Health.cpp:
```diff
@@ -0,0 +1,75 @@
+    float wind_max = AP::ahrs().get_max_wind();
+    uint32_t current_time = AP_HAL::millis();
+    uint32_t time_since_last_call =  current_time - state[i].airspeed_failure_last_check;
+    if(time_since_last_call > AP_AIRSPEED_FAILURE_TIME_PERIOD_MS) {
```
time_since_last_call doesn't need it's own variable. It's used twice. Just use:
` (current_time - state[i].airspeed_failure_last_check > AP_AIRSPEED_FAILURE_TIME_PERIOD_MS)`
like everywhere else in the codebase. Same goes for the one below. Also, if these are all in millisec, then please name then with _ms like all the other millisecond variables in the codebase
rename current_time to now_ms and state[i].airspeed_failure_last_check to state[i].failure.last_check_ms

magicrub commented on libraries/AP_Airspeed/AP_Airspeed_Health.cpp:
```diff
@@ -0,0 +1,75 @@
+                float gnd_speed = AP::gps().ground_speed();
+
+                if(aspeed < (gnd_speed + wind_max) && aspeed > (gnd_speed - wind_max)) {
+                    state[i].airspeed_failure_basic_success_count++;
```
Again, the count is related to sample rate. What you're actually looking for is one second of consistently good samples and any bad sample will reset the one second timer. So, timestamp the last bad sample and if you ever see a bad sample set it to now_ms and if (now_ms - last_bad_ms < 1000) then you know you're in the bad state. Right now yo have timers setting counters. Just use timers!

ChrisBird replied:
Are you referring to just the enabling portion?  If so i can agree with that but over a longer period.

If your also referring to the initiating the failure part i disagree.  There are times where it drops say as a turn is made into the wind, or short gust catches it.  If you have an alternative where i can do that without counters of some kind that would be great.  A single bad reading shouldnt cause a failure.

ChrisBird replied:
I'll use the low pass filter option, much better solution

ChrisBird commented on libraries/AP_AHRS/AP_AHRS.h:
```diff
@@ -304,6 +304,10 @@ class AP_AHRS
         return _airspeed != nullptr && _airspeed->use() && _airspeed->healthy();
     }
 
+    uint8_t get_max_wind() const {
```
Sure no probs

ChrisBird commented on libraries/AP_Airspeed/AP_Airspeed.cpp:
```diff
@@ -129,6 +129,13 @@ const AP_Param::GroupInfo AP_Airspeed::var_info[] = {
+    // @Description: Bitmask of what options to use with airspeed.
+    // @Bitmask: 0:Use Basic Airspeed Failure Check
+    // @User: Advanced
+    AP_GROUPINFO("_OPTIONS",  21, AP_Airspeed, _options, AP_AIRSPEED_OPTIONS_DEFAULT),
```
Sure no probs.  Is it better for it to be moved to the bottom?

ChrisBird commented on libraries/AP_Airspeed/AP_Airspeed.h:
```diff
@@ -175,6 +187,7 @@ class AP_Airspeed
     static AP_Airspeed *_singleton;
 
     AP_Int8 primary_sensor;
+    AP_Int16    _options;    // bitmask options for airspeed
```
Sure no probs

ChrisBird commented:
@magicrub  I think I've applied most if not all your comments.  Thanks for the suggestion of the low pass filter it has worked even better than I had thought.

I rejigged the code so that it does the filtering first, then does the handling of the filtered value.  The filter enabled me to get rid of not only the counter variables but I rethought some of the logic which allowed me to further reduce them.

I've gone with an asymmetric variation on the filter so that the good readings impact the filtered value less.  I've also added a warning level in so that if a person has a plane and they are out on their max airspeed estimate by a small amount they may have time to fix it.  I ran a quite a few SITL tests with various wind speeds, airspeed failures and turbulence levels.  It handled exactly as I would expect.

I have far more confidence in the auto enable feature with the filtered value, it is very unlikely to turn on during the delicate landing phase unless the issue is cleared.   It will take about 5 secs to trip a failure if they are all bad readings.  It takes about 45 to 60 seconds to return to an ok state.

Damn it, I just realised I left the default option levels at 3 which is all on.  I'll fix and repush it.......

magicrub commented:
Hi @ChrisBird , thanks for working on this! It's getting much better. I re-wrote a few sections of it at pushed it to my repo branch "[pitot_failsafe](https://github.com/magicrub/ardupilot/tree/pitot_failsafe)". Can you please look over this commit and test it? 

https://github.com/magicrub/ardupilot/commit/65963058210ae11bf9c7e030856590ddd12cf1d0

general changes:
- removed the dedicated file, it's just implementing one function. moved it ot the main cpp
- renamed a few things
- used a LPF properly for the probability (actually stays between 0 and 1)
- improved the warning trigger to only happen once, and ensure it always happens
- saved 200 bytes!
- exit sooner if there's no work to do
- sprinkled in some cost
- improved the bounds checking of .use

ChrisBird commented:
Hi Rob,

The additional file was requested on the previous dev call by Tridge as we'll be adding to the health checks in the near future.

I'll have a look tonight and get back to you.

magicrub commented:
I re-added the file on another commit on my branch.

ChrisBird commented:
@magicrub, I've had a quick look.  I still think some of the changes are not setting up the frame work for the future checks.  The naming of enum items seem to support just one health check.  I would envisage that there would eventually be different checks that people will be able to turn on, not just one health check.  The KF version may not suite some people going forwards and they'd refer the simplistic check, or maybe they want both enabled.  Not all will want the simple check once the KF version is done.  I agree with most of the other changes, I'm going to test it now.  What we have on by default will change as newer or better health checks come out.  The airspeed health seems to have alot of different views and given that people appear to feel like this hasnt been addressed in years I think we should try to placate people a bit here.  Of course once the more robust version is out then others will hopefully trust it more than the simple check.

magicrub commented:
per @ChrisBird 's request via gitter, he liked my version and was ok with me force-pushing it onto his branch.

ChrisBird commented:
We can remove the WIP tag.  I think this PR is looking good.

magicrub commented:
I reluctantly agree to merge this as-is but let me state my case of why I prefer the other way.

The options check is better done inside check_sensor_failures() with it returning immediately if both are not enabled. This allows the same functionality but the AP_Airspeed::OptionsMask::ON_FAILURE_ defines would only ever be called inside check_sensor_ahrs_wind_max_failures() which makes it more obvious that it's use is confined there. Plus, if anyone else calls check_sensor_ahrs_wind_max_failures() in the future it ensures it's properly bypassed without doing unnecessary work and same behavior. For example, if someone else called this function in the future from somewhere else then probability would update where you version disables that update. Future coders are worthwhile to protect against :)

Also of note, my original rewrite did not check options until it used it because always updating the probability is useful. It probability should get logged too (new PR).

Naterater commented:
The whole thesis of this thread is that we should be allowing switch over to throttle mode in the event of airspeed failure. Should we be checking the sanity TRIM_THROTTLE during level cruise when the airspeed is healthy? That could be a potential future failsafe, but for now at least a warning would be quite easy to warn on the GCS "Caution: TRIM_THROTTLE exceeded in automatic level flight." It would be an indicator that AHRS_WIND_MAX is about to be exceeded if the wind isn't truly that high. Maybe this is just a feature to add to a future PR with more advanced failsafes, but I wanted to document the idea so I don't forget.

magicrub commented:
I've adding logging of the probability which is only useful when it's always being calculated so I went back to always running it, and only perform the actions if options says so. We can rework that later as we get more options if needed since we may be having different probabilities for different algorithms. 

Meanwhile, it would be nice to see some real flight logs with the health probability with or without the feature enabled.

ChrisBird commented:
Hi @magicrub, the reason why I think it shoudn't calculate it is say we have another checking method, say one that is working of a synthetic airspeed (like we have but use it to determine if its a solid reading or not).  I'd like to use the same probability parameter, if I have the option off then I dont want it to modify the probability value, I only want those options that I have on to modify the probability.

So in the next go with the synthetic airspeed I'll split them out into two separate methods so that way they are decoupled, further.  I'd see us having one method that looks at the probability and turns it on or off and then multiple method that may modify the probability.  Does that make sense?  Or should I approach this differently?

magicrub commented:
Yeah, best to separate the probability as a stand alone calc. At some point it may be useful to have multiple probabilities. If so, we will need to make a new logging msg.
